File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void helper(vector<int>& nums, vector<int>input, int n, vector<vector<int>>&res, int prev){
        
        if(n == 0){
            res.push_back(input);
            return;
        }
        
        while(n > 0 && nums[n - 1] == prev)     
            --n;
        
        if(n == 0)  return;
        
        input.push_back(nums[n - 1]);
        
        helper(nums, input, n - 1, res, nums[n - 1]);
        
        input.pop_back();
        
        helper(nums, input, n - 1, res, prev);
        
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        
        sort(nums.begin(), nums.end());
        
        vector<vector<int>>res;
        
        vector<int>input;
        
        int n = nums.size();
        
        helper(nums, input, n, res, -100);
        
        return res;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    private List<List<Integer>> res;
    
    private void helper(int[] nums, int idx, List<Integer>smallRes){
        if(idx == nums.length){
            res.add(new ArrayList<>(smallRes));
            return;
        }
        
        smallRes.add(nums[idx]);
        helper(nums, idx + 1, smallRes);
        smallRes.remove(smallRes.size() - 1);
        
        helper(nums, idx + 1, smallRes);
    }
    
    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        List<Integer>smallRes = new ArrayList<>();
        
        helper(nums, 0, smallRes);
        
        return res;
    }
}